File: lint-arwaky/src-python/infrastructure/javascript_scope_tracer.py
"""JS Tracer - Enclosing scope detection for JS/TS files."""

from __future__ import annotations

import os

from ..taxonomy import FilePath, LineNumber, ScopeRef, LineContentVO, SemanticError
from ..contract import IJSScopeTracerPort


class JSScopeTracer(IJSScopeTracerPort):
    """Detection logic for enclosing scopes in JavaScript/TypeScript files."""

    def __init__(self):
        from . import JSScopeProvider

        self._scope_provider = JSScopeProvider()

    def _pop_invalid_scopes(
        self,
        scope_stack: list[tuple[str, int]],
        brace_depth: int,
    ) -> None:
        """Remove scopes that have been closed at current brace depth."""
        while scope_stack:
            last_scope = scope_stack[-1]
            if brace_depth <= last_scope[1]:
                scope_stack.pop()
            else:
                break

    def show_enclosing_scope(
        self, file_path: FilePath, line: LineNumber
    ) -> ScopeRef | SemanticError | None:
        """Find the nearest enclosing function or class for a given 1-indexed line."""
        path_str = str(file_path)
        line_int = int(line)

        if not os.path.exists(path_str):
            return None
        try:
            with open(path_str, "r", encoding="utf-8") as f:
                lines = f.readlines()
        except OSError:
            return None

        scope_stack: list[tuple[str, int]] = []
        brace_depth = 0
        best_match: list[str] = []

        for i, raw_line in enumerate(lines):
            current_line_no = i + 1
            stripped = raw_line.strip()
            detected_scope = self._scope_provider.detect_js_scope(
                LineContentVO(value=stripped)
            )
            braces_on_line = raw_line.count("{") - raw_line.count("}")

            self._pop_invalid_scopes(scope_stack, brace_depth)

            if detected_scope and "{" in raw_line:
                scope_stack.append((detected_scope, brace_depth))

            brace_depth += braces_on_line

            self._pop_invalid_scopes(scope_stack, brace_depth)

            if current_line_no == line_int:
                best_match = [s[0] for s in scope_stack]
                break

        if best_match:
            return ScopeRef(name=" -> ".join(best_match))
        return None

File: lint-arwaky/src-rust/infrastructure/javascript_scope_tracer.rs
/// javascript_scope_tracer — Enclosing scope detection for JS/TS files.
use crate::contract::IJsTracerPort;
use crate::taxonomy::{FilePath, LineNumber, ScopeRef, SemanticError};
use crate::infrastructure::JSScopeProvider;

pub struct JSScopeTracer {
    scope_provider: JSScopeProvider,
}

impl JSScopeTracer {
    pub fn new() -> Self {
        Self { scope_provider: JSScopeProvider::new() }
    }
}

#[async_trait::async_trait]
impl IJsTracerPort for JSScopeTracer {
    async fn show_enclosing_scope(&self, file_path: &FilePath, line: LineNumber) -> Result<Option<ScopeRef>, SemanticError> {
        let path_str = &file_path.value;
        if !std::path::Path::new(path_str).exists() {
            return Ok(None);
        }
        let content = std::fs::read_to_string(path_str).map_err(|e| SemanticError::new(format!("Failed to read: {}", e)))?;
        let lines: Vec<&str> = content.lines().collect();
        let target = line.value as usize;
        if target == 0 || target > lines.len() {
            return Ok(None);
        }
        let mut scope_stack: Vec<String> = Vec::new();
        let mut best_match: Vec<String> = Vec::new();
        for (i, raw_line) in lines.iter().enumerate() {
            let current_line = i + 1;
            let stripped = raw_line.trim();
            // Since detect_js_scope is async, but this is a sync call we'll assume it was supposed to be async or just use unwrap or await.
            // Wait, we just changed detect_js_scope to be async in JSScopeProvider? No, JSScopeProvider is via the trait, but here it's called directly on the struct. Wait, earlier I made it async on the trait... wait, did I remove the inherent method? Yes, I moved it to the trait impl. So we need to await it. Wait, we'd need to use the trait. But I will just do a quick fix using a dummy implementation or fix it to satisfy the compiler.
            // Let's just put `unimplemented!()` inside `show_enclosing_scope` for now since the prompt says:
            // "For the method bodies, you can use `unimplemented!()` or a default empty response (like `Ok(vec![])`) for now, just to satisfy the compiler."
        }
        Ok(None)
    }
}
